crates/hir-ty/src/infer/unify.rs

```diff
@@ -146,6 +147,8 @@ pub(crate) struct InferenceTable<'a> {
     /// Double buffer used in [`Self::resolve_obligations_as_possible`] to cut down on
     /// temporary allocations.
     resolve_obligations_buffer: Vec<Canonicalized<InEnvironment<Goal>>>,
+    fn_trait_predicates: Vec<(Ty, FnTrait)>,
+    cached_fn_trait_ids: Option<CachedFnTraitIds>,
```

Comment:
Not too happy with this, can't we just iterate through the pending obligations when necessary? (as is what rustc does)

Reply:
The chalk resolves obligations like `FnOnce` immediately, so they are not remained in the pending obligations. The most problematic thing is that it infers the closures with `FnOnce` predicate as implementing `FnMut` and `Fn`

Reply:
Maybe I should look into it more, but I'm not sure that it is intended chalk functionality or not

Reply:
Anyway, I'm gonna take a deep look into the chalk closure inferences. I might fix this from upstream or maybe RA is asking chalk without some details

Reply:
@Veykril I had a closer into RA and chalk, and I think that we couldn't use pending obligations like rustc does.

When the chalk has to solve a goal that whether some closure implements some `FnTrait`;

[1. It asks `RustIrDatabase` for the kind of our closure and checkes if the kind of closure is compatible with the given `FnTrait`](https://github.com/rust-lang/chalk/blob/3b710e1be212912610b0ad4fcefd7f8306f1c33b/chalk-solve/src/clauses/builtin_traits/fn_family.rs#L110-L130)
[2. If so, it pushes the clause for the "FACT" that our closure implements that trait.](https://github.com/rust-lang/chalk/blob/3b710e1be212912610b0ad4fcefd7f8306f1c33b/chalk-solve/src/clauses/builtin_traits/fn_family.rs#L9-L25)

So, solving whether the given closure implements `FnTrait` is heavily depends on `RustIrDatabase` and I think that 's reasonable for it to do so because what chalks wants to be(though it is not going to be actively maintained due to new trait solver in rustc) is [an abstract trait solver](https://rust-lang.github.io/chalk/book/types.html#goal-of-the-chalk-ir-crate), not a very definite rustc inference maker.

But since RA's `ChalkContext`, which is our `RustIrDatabase` implementation, [always returns `Fn` as the kind of a closure](https://github.com/rust-lang/rust-analyzer/blob/a01655552df5a4f7ef2db8296d3ad15746faa22c/crates/hir-ty/src/chalk_db.rs#L359-L366). (I think that this is sort of a heuristic to avoid having other inferences and coercions delayed or unresolved.)
Thus, the obligations like `closure{n}[...] implements FnX` are always solved into `Unique` solutions and not remain as "pending" because they are immediately solved.

Giving the "correct" kind of a closure doesn't work for this because;

1. We need to use "correct" pending obligations for closure kind inferencing
2. But the "correct" pending obligations are what we are going to do with "correct" kind of a closure, a cyclic dependency!

I've tried just giving `FnOnce` as a `closure_kind` but aside from that it makes lots of inferences in our test cases unresolved - if the later were to worked correctly I would have tried fixing those inferences -, it still does not "pend" obligations.
The obligations for closure implementation `FnMut` and `Fn` are solved as `None`, and [they are not pended as well](https://github.com/rust-lang/rust-analyzer/blob/c864953c0f1feeed332c08b8c17192eedbc2535f/crates/hir-ty/src/infer/unify.rs#L675-L707).

So, I have concluded that we would have to modify lots of RA inferencing code to do this as rustc does, and this was why I wrote somewhat sad codes for similar thing 😢  (When I wrote it, I hadn't debugged the chalk much, but tried using pending obligations and it was always empty; so I had similar conclusion without much evidence then 😅 )